### PR TITLE
[Feature] Extend tutor leaderboards with average_score and average_rating

### PR DESCRIPTION
**Pull request by Olimbia23**
### Checklist
- [X] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [X] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [X] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [ ] Server*: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [X] Server: I implemented the changes with a good performance and prevented too many database calls
- [X] Server: I documented the Java code using JavaDoc style.
- [X] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [ ] Client*: I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-testing/).
- [ ] Client*: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [X] Client: I documented the TypeScript code using JSDoc style.
- [X] Client: I added multiple screenshots/screencasts of my UI changes
- [X] Client: I translated all newly inserted strings into English and German.
- [X] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).

[ ]* : not needed. 

### Motivation and Context

Tutors' Leaderboards are an important part of Artemis which show information about the performance of the tutors. 
There are 2 types of leaderboards: Course and Exercise leaderboards. 
The motivation for this PR is to **enrich the Tutor Leaderboards**, both for course and exercise, with **2 new columns** named **Average Score** and **Average Ratings**.  
The _average score_ indicates the average of all the scores that a tutor has given when reviewing students’ submission.
While the _average rating_ is the value that indicates how many stars in average(1-5) has a tutor received by the students.

### Description

I extended `TutorLeaderboardDTO` to contain `averageScore` and `averageRating` fields. 
`averageScore` and `averageRating` fields are computed dynamically by the database on the fly during the request. Extended the table to have two extra columns with some formatting of the double values. Added two integration tests that test the average computation on the server.

### Steps for Testing

1. Login to Artemis as an Instructor and create 1 exercise(any type). set the due date and assessment date with ~5 difference so you can have time to execute steps 2-3
2. Login to Artemis as (at least 2) different students, solve the exercise and submit.
3. Login to Artemis as a tutor and review both of the 2 student’s submissions. (Remember the tutor's name)
4. Login to Artemis with each student account and rate the tutor (this step is necessary to test the ratings)
5. Now, you can login to Artemis as a tutor/instructor and go to the Exercise Assessment Dashboard.
6. Scroll down until you reach the Tutor Exercise Leaderboard.
7. There should be 2 new columns: Average Score and Average Rating.
8. Verify with your calculator if the values displayed in the table are correct.
9. Don't forget to also check the values in the Tutor Course Leaderboard.

### Review Progress

- Code Review
  - [x] Review 1
  - [x] Review 2
- Manual Tests
  - [x] Test 1
  - [x] Test 2

### Test Coverage

TutorLeaderboardService.java: 100%

### Screenshots

Screenshot 2021-09-20 at 8 12 29 PM

- - -

Screenshot 2021-09-20 at 8 10 45 PM

### CODE REVIEW

**Review by Malyuk-A — COMMENTED** [review]
Code looks good.
Tested on TS3. The feature works as expected 👍

But I noticed some problematic behavior. I do not know if this is due to this PR (database change) or due to the update PR.

Image 1 : After assesing I clicked the Save button and the buttons were stuck (infinite loading loop). There seems to be an error with sending statistics (maybe @ndricimrr might know more).
image

Image 2 : after reloading and clicking on submit it seems to work but the debug console showed errors with the statistics. 
image

File: src/test/java/de/tum/in/www1/artemis/TutorLeaderboardServiceIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+public class TutorLeaderboardServiceIntegrationTest extends AbstractSpringIntegrationBambooBitbucketJiraTest {
+
+    @Autowired
+    private TutorLeaderboardService tutorLeaderboardService;
+
+    final private static int TUTOR_COUNT = 1;
+
+    final private static int ASSESSMENT_COUNT = 2;
+
+    Course course;
```
Comment (imstreetspirit) [inline src/test/java/de/tum/in/www1/artemis/TutorLeaderboardServiceIntegrationTest.java:31]:
```suggestion
    private Course course;
```

File: src/test/java/de/tum/in/www1/artemis/TutorLeaderboardServiceIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    Exercise exercise;
```
Comment (imstreetspirit) [inline src/test/java/de/tum/in/www1/artemis/TutorLeaderboardServiceIntegrationTest.java:33]:
```suggestion
    private Exercise exercise;
```

File: src/test/java/de/tum/in/www1/artemis/TutorLeaderboardServiceIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    /**
+     * Prepares the testing suite by initializing variables and mocks
+     */
+    @BeforeEach
+    public void reset() {
```
Comment (imstreetspirit) [inline src/test/java/de/tum/in/www1/artemis/TutorLeaderboardServiceIntegrationTest.java:39]: I just added this suggestion to keep consistency between integration test classes
```suggestion
    public void initTestCase() {
```

**Review by imstreetspirit — COMMENTED** [review]
I just left very small suggestions, besides that looks good to me.

**Review by lschlesinger — COMMENTED** [review]
The code looks good, i did not find anything to improve. The information is a good addition to the table, I was just wondering, if it is immediately clear, what these values mean. Especially since there are not points, score and rating... it could be confusing. I'd say that "Avg. given score" and "Avg. received rating" would be more descriptive, but I understand that these labels would be too long. Maybe the `?` icon with a tooltip would be an idea? May it's clear to the users and it's just me not being a regular user of this feature.
Also, I think the average score could be placed next to points.
Furthermore, I was wondering if there are not client tests that need adaption? I guess not, otherwise you would have done it or the existing ones would fail, but it looks to be as if the tests for that component need some improvement though, i.e. testing for the columns etc.?

I as well tested the functionality on TS1, assessed a modeling and a text exercise (had the same issue as @Malyuk-A, it's not related to you PR, pretty sure). Both tables were updated correctly. Let me know if you want to change something on the table labels (order, name), then I'll approve the PR :)

**Comment by Olimbia23** [comment]
Thank you @imstreetspirit. I applied your suggestions 😊

**Comment by Olimbia23** [comment]
Thank you for testing the PR and also finding the bug @Malyuk-A 😇 
Even though it has nothing to do with this PR I created a fix for it: https://github.com/ls1intum/Artemis/pull/4039

**Comment by Olimbia23** [comment]
Thank you for reviewing and testing it @lschlesinger 👼 

I think these two columns have to be together due to their similarity of having the average prefix, it helps with readability.

I think for regular users the naming of the columns is clear.

Since you also noticed the bug that @Malyuk-A found, could you please test if I have properly fixed it in this PR: https://github.com/ls1intum/Artemis/pull/4039

For me it was working for modelling exercises even without this fix.

**Review by lschlesinger — APPROVED** [review]
Thanks for clarification, it's fine for me.
Btw. the bug only appears for text exercises, I tested your PR and left the comment there.

**Review by ndricimrr — COMMENTED** [review]
Code changes look good. 

Wasn't able to test on test servers, but tested locally. 

I had an issue with opening the Exercise Dashboard page as an instructor/admin. I get an infinitely looping client error. Looks something like this: 
<details>

test001



zzzz01

**Comment by Olimbia23** [comment]
Thank you for testing it locally @ndricimrr.

Are you sure that you were running not only the client of my branch PR, but also the server of it as it passes new data to the client.
The build is green again. Could you maybe check it on the test-server and let me know if you have this issue again.

**Review by ndricimrr — APPROVED** [review]
Tested again on TS2. Worked as described. Apparently some issue with my local setup.

**Review by imstreetspirit — APPROVED** [review]
looked at changes after my last review. my points were very small thing, but you already addressed them, thank you. looks good to me

**Pull request merged**